added btc convertion tools to utils

jonathansmirnoff opened the pull request:
It was added some functions that are used in https://github.com/rsksmart/utils/ to rsk3 utils.
Please @josedahlquist, @bguiz and @alebanzas check also this.

jonathansmirnoff committed:
added btc convertion tools to utils

bguiz commented on packages/rsk3-utils/src/index.js:
```diff
@@ -1046,6 +1050,81 @@ const soliditySha3 = function() {
+ *
+ * @return {String} RSK Private Key.
+ */
+function getRskPrivateKey(btcPrivateKey) {
```
```suggestion
const getRskPrivateKey = function (btcPrivateKey) {
```

for consistency with the other functions

bguiz commented on packages/rsk3-utils/src/index.js:
```diff
@@ -1046,6 +1050,81 @@ const soliditySha3 = function() {
+    partialResults.push(0x01);
+    var check = convertHex.hexToBytes(sha256(convertHex.hexToBytes(sha256(partialResults))));
+
+    for (var partialResul of partialResults) {
+        result.push(partialResul);
+    }
```
```suggestion
    for (var partialResult of partialResults) {
        result.push(partialResult);
    }
```

bguiz commented on packages/rsk3-utils/tests/utils.test.js:
```diff
@@ -691,4 +694,44 @@ describe('UtilsTest', () => {
         expect(() => jsonInterfaceMethodToString({name: 'name'})).toThrow();
     });
+
+    it('calls to getAddress and returns de expected value', () => {
```
```suggestion
    it('calls getAddress and returns the expected value', () => {
```

bguiz replied:
^ bump - proofreading edit

bguiz commented on packages/rsk3-utils/tests/utils.test.js:
```diff
@@ -691,4 +694,44 @@ describe('UtilsTest', () => {
+        });
+    });
+
+    it('calls to getRskPrivateKey and returns de expected value', () => {
```
```suggestion
    it('calls getRskPrivateKey and returns the expected value', () => {
```

bguiz commented on packages/rsk3-utils/tests/utils.test.js:
```diff
@@ -691,4 +694,44 @@ describe('UtilsTest', () => {
+        });
+    });
+
+    it('calls to getBtcPrivateKey and returns de expected btc private key', () => {
```
```suggestion
    it('calls getBtcPrivateKey and returns the expected btc private key', () => {
```

bguiz commented on packages/rsk3-utils/tests/utils.test.js:
```diff
@@ -691,4 +694,44 @@ describe('UtilsTest', () => {
     });
+
+    it('calls to getAddress and returns de expected value', () => {
+        const tests = [
+            {
+                value: 'Kxj9x4G2Uvw8CbngATxGZLY4K2EFkjty7mumLabJQEhHw9ZLyZTs',
+                expected: '36eb6a1d1720a6037d0680b8b357fc661a24bfdc'
+            }
+        ];
```
Having an array of tests and then looping over them seems redundant if there is only going to be one test case.

To improve this I suggest having at least one "happy path" test case, which is what you already have here, and at least one "failure path" test case

For the failure path test cases, the function being called is expected to error. For example, if your provide it with an invalid input `value`.

In the implementations for these functions being tested, I currently do not see any validation logic, so perhaps the inclusion of these test cases will make for a more robust implementation.

bguiz commented on packages/rsk3-utils/tests/utils.test.js:
```diff
@@ -691,4 +694,44 @@ describe('UtilsTest', () => {
+            }
+        ];
+
+        tests.forEach((test) => {
+            expect(getRskPrivateKey(test.value)).toEqual(test.expected);
+        });
```
Same comment as above regarding including test cases for "failure paths".

bguiz commented on packages/rsk3-utils/tests/utils.test.js:
```diff
@@ -691,4 +694,44 @@ describe('UtilsTest', () => {
+            }
+        ];
+
+        tests.forEach((test) => {
+            expect(getBtcPrivateKey(test.mainnetNetwork, test.value)).toEqual(test.expected);
+        });
```
Same comment as above regarding including test cases for "failure paths".

bguiz commented on packages/rsk3-utils/src/index.js:
```diff
@@ -1046,6 +1050,81 @@ const soliditySha3 = function() {
+ * @return {String} RSK Private Key.
+ */
+function getRskPrivateKey(btcPrivateKey) {
+    var privKeyBytes = _keyBtcToRskInBytes(btcPrivateKey);
```
before performing the conversion, validate that the input value, `btcPrivateKey`.

bguiz commented on packages/rsk3-utils/src/index.js:
```diff
@@ -1046,6 +1050,81 @@ const soliditySha3 = function() {
+}
+
+const _keyBtcToRskInBytes = function(privKeyAsExportedByBitcoinDumpprivkey) {
+    var decodedKey = bs58.decode(privKeyAsExportedByBitcoinDumpprivkey);
```
before performing the conversion, validate that the input value, `privKeyAsExportedByBitcoinDumpprivkey`.

bguiz commented on packages/rsk3-utils/src/index.js:
```diff
@@ -1046,6 +1050,81 @@ const soliditySha3 = function() {
+ */
+
+const getBtcPrivateKey = function(isBtcMainnet, rskPrivateKey) {
+    var addressArray = convertHex.hexToBytes(rskPrivateKey);
```
before performing the conversion, validate that the input value, `rskPrivateKey`.

jonathansmirnoff commented:
@bguiz I updated the PR with your suggestions. For the validations I added bitcoinjs-lib.
Now we have btc private key validation and also rsk private key validation.

Also please @josedahlquist take a look of this when you can.

alepc253 reviewed (approved):
LGTM. The `null` return in case of invalid pk could be improved.

jonathansmirnoff commented:
> LGTM. The `null` return in case of invalid pk could be improved.

Yes, totally agree. We should think in a standard way. @chrisli30 what do you think? What is your recommendation?
Are you OK to add here bitcoinjs-lib and ethereumjs-wallet?

bguiz commented on packages/rsk3-utils/src/index.js:
```diff
@@ -1046,6 +1051,123 @@ const soliditySha3 = function() {
+ * @return {String} RSK address.
+ */
+const getRskAddress = function(btcPrivateKey) {
+    if (!isValidBtcPrivateKey(btcPrivateKey)) return null;
```
Regarding @alepc253 's comment, rather than return `null`, a viable alternative would be to throw an error. For example:

```suggestion
    if (!isValidBtcPrivateKey(btcPrivateKey)) {
      throw new Error('Invalid BTC private key');
    }
```

(If you change this here, change it in the other functions too.)

The impact of this would be slightly more complex code within the calling function - it would need to use a try...catch block instead of an if block. That being said, both approaches are valid, and I would decide - for the sake of consistency - based on the convention used within the rest of rsk3.js. Let's wait for @chrisli30 to weigh in.

bguiz commented on packages/rsk3-utils/tests/utils.test.js:
```diff
@@ -691,4 +696,193 @@ describe('UtilsTest', () => {
     });
+
+    it('calls to getAddress and returns de expected value', () => {
+        const tests = [
+            {
+                value: 'Kxj9x4G2Uvw8CbngATxGZLY4K2EFkjty7mumLabJQEhHw9ZLyZTs',
+                expected: '36eb6a1d1720a6037d0680b8b357fc661a24bfdc'
+            },
+            {
+                value: 'L5eLLCcuBK8WGsSNcm8gBGp8cZUzsCviVF2DuPeiCaPpH1aiEmhE',
+                expected: '3f6ed7d236ccdb6b4cb07018793a206f9c1f93cc'
+            },
+            {
+                value: 'L5EZS1X7DYHjZNH8C44xZVUHRgHcJ3SjugmmArBuKDv1aTWfs878',
+                expected: '138811cdf0544996d39fc41fc2a5d428392ab123'
+            },
+            {
+                value: '',
+                expected: null
+            },
+            {
+                value: null,
+                expected: null
+            },
+            {
+                value: 1,
+                expected: null
+            },
+            {
+                value: 'Ky4fD7LXsQ!!!!!!!QwTKq91g1LFFbRpFhuW4NeBJxn73Do2cSuze1NWcLY',
+                expected: null
+            }
+        ];
```
:+1: thanks for making this change - the tests feel a lot more solid now!

bguiz commented on packages/rsk3-utils/src/index.js:
```diff
@@ -1046,6 +1051,123 @@ const soliditySha3 = function() {
+        const privateKeyBuffer = Buffer.from(rskPrivateKey, 'hex');
+        wallet.fromPrivateKey(privateKeyBuffer);
+    } catch (error) {
+        console.log(error);
```
```suggestion
```

(let's remove any debugging output)

jonathansmirnoff commented:
@chrisli30 please check this PR

chrisli30 reviewed (commented):
Hi guys, I have looked at this changelist and have no problem with the code, but serious security concerns were raised on those conversion functions during our external security audit with Quantstamp. There's risk for rsk3.js library to include them. I know this is a nice utility and we included them in the early version of rsk3.js, but raising a security red flag here 🟥

jonathansmirnoff commented:
@jose

> Hi guys, I have looked at this changelist and have no problem with the code, but serious security concerns were raised on those conversion functions during our external security audit with Quantstamp. There's risk for rsk3.js library to include them. I know this is a nice utility and we included them in the early version of rsk3.js, but raising a security red flag here 🟥

Thanks for the explanation. If we add this changes to rsk-util it will be also added in rsk3.js?

chrisli30 commented:
> @jose
> 
> > Hi guys, I have looked at this changelist and have no problem with the code, but serious security concerns were raised on those conversion functions during our external security audit with Quantstamp. There's risk for rsk3.js library to include them. I know this is a nice utility and we included them in the early version of rsk3.js, but raising a security red flag here 🟥
> 
> Thanks for the explanation. If we add this changes to rsk-util it will be also added in rsk3.js?

Yes, Jon. The rsk3.js is packaged together using lerna multi-package management. When user do npm install @rsksmart/rsk3.js  they will install the rsk3-utils too.

Do you think we can make the conversion utility another small library, and add specific alerts to the users?

It doesn't have to be all in one library in my opinion, and the main purpose of rsk3.js is to manage RBTC and RIF wallets.

jonathansmirnoff commented:
Thanks Chris for your feedback. I am going to create another library with just this logic to avoid compromise rsk3.js.
